ui/src/app/utils/authentication.tsx

```diff
@@ -0,0 +1,92 @@
+import {AnalyticsTracker, setLoggedInState} from 'app/utils/analytics';
+import {LOCAL_STORAGE_KEY_TEST_ACCESS_TOKEN} from 'app/utils/cookies';
+import {authStore, serverConfigStore, useStore} from 'app/utils/stores';
+import {environment} from 'environments/environment';
+import {ConfigResponse} from 'generated/fetch';
+import {useEffect} from 'react';
+
+declare const gapi: any;
+
+// for e2e tests: provide your own oauth token to obviate Google's oauth UI
+// flow, thereby avoiding inevitable challenges as Google identifies Puppeteer
+// as non-human.
+declare global {
+  interface Window { setTestAccessTokenOverride: (token: string) => void; }
+}
+
+const makeAuth2 = (config: ConfigResponse): Promise<any> => {
+  return new Promise((resolve) => {
+    gapi.load('auth2', () => {
+      gapi.auth2.init({
+        client_id: environment.clientId,
+        hosted_domain: config.gsuiteDomain,
+        scope: 'https://www.googleapis.com/auth/plus.login openid profile'
+            + (config.enableBillingUpgrade ? ' https://www.googleapis.com/auth/cloud-billing' : '')
+      }).then(() => {
+        authStore.set({
+          ...authStore.get(),
+          authLoaded: true,
+          isSignedIn: gapi.auth2.getAuthInstance().isSignedIn.get()
+        });
+
+        gapi.auth2.getAuthInstance().isSignedIn.listen((nextIsSignedIn: boolean) => {
+          authStore.set({...authStore.get(), isSignedIn: nextIsSignedIn});
+        });
+      });
+      resolve(gapi.auth2);
+    });
+  });
+};
+
+export const signIn = (): void => {
```

Comment:
this requires that gapi.auth2 is already loaded - would either comment or somehow block on that here

Reply:
we could just load it again? I think subsequent loads of the same module are fast.